quickwit-directories/src/bundle_storage.rs

```diff
@@ -0,0 +1,289 @@
+        path: &Path,
+        _output_path: &Path,
+    ) -> quickwit_storage::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn get_slice(
+        &self,
+        path: &Path,
+        range: Range<usize>,
+    ) -> quickwit_storage::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        let new_range =
+            range.start + file_offsets.start as usize..range.end + file_offsets.start as usize;
+        self.storage
+            .get_slice(&self.bundle_file_name, new_range)
+            .await
+    }
+
+    async fn get_all(&self, path: &Path) -> quickwit_storage::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        self.storage
+            .get_slice(
+                &self.bundle_file_name,
+                file_offsets.start as usize..file_offsets.end as usize,
+            )
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> quickwit_storage::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn exists(&self, path: &Path) -> quickwit_storage::StorageResult<bool> {
+        // also check if self.bundle_file_name exists ?
+        Ok(self.metadata.exists(path))
+    }
+
+    fn uri(&self) -> String {
+        self.storage.uri()
+    }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+struct FileOffsets {
+    start: u64,
+    end: u64,
+}
+
+impl HasLen for BundleStorage {
+    fn len(&self) -> usize {
+        unimplemented!()
+    }
+}
+
+impl fmt::Debug for BundleStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BundleStorage({:?}, files={:?})",
+            &self.bundle_file_name, self.metadata
+        )
+    }
+}
+
+fn unsupported_operation(path: &Path) -> StorageError {
+    let msg = "Unsupported operation. BundleStorage only supports async reads";
+    error!(path=?path, msg);
+    io::Error::new(io::ErrorKind::Other, format!("{}: {:?}", msg, path)).into()
+}
+
+/// BundleStorage bundles together multiple files into a single file
+/// with some metadata
+pub struct CreateBundleStorage {
+    metadata: BundleStorageMetadata,
+    current_offset: u64,
+    /// the offset in the file where the hotcache begins. This is used to read
+    /// the hotcache and footer in a single read.
+    pub hotcache_offset: u64,
+    bundle_file: File,
+}
+
+impl CreateBundleStorage {
+    /// Creates a new CreateBundleStorage, to which files can be added.
+    pub fn new(path: &Path) -> io::Result<Self> {
+        let sink = OpenOptions::new().create(true).append(true).open(path)?;
+
+        Ok(CreateBundleStorage {
+            bundle_file: sink,
+            current_offset: 0,
+            hotcache_offset: 0,
+            metadata: Default::default(),
+        })
+    }
+
+    /// Appends a file to the bundle file.
+    ///
+    /// The hotcache needs to be the last file that is added, in order to be able to read
```

Comment:
I'm worried about the user that will add the hotcache at the beginning. 

What will happen if add it at the beginning and even if we forget to add the hotcache? Can we protect us against these two situations?

Reply:
I added some checks, but did not yet commit them